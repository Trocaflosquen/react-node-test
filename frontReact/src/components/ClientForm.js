import React, { Component } from 'react'
import { Glyphicon, FormControl, ControlLabel, FormGroup, Button } from 'react-bootstrap'
import ClientElement from './ClientElement';


class ClientForm extends Component {

    constructor(props) { 
        super(props)
        this.handleChange = this.handleChange.bind(this)
        this.state = {
            name: '',
            phone: '',
            username: '',
            website: '',
            email: '',
            company: {
                bs: '',
                catchPhrase: '',
                name: ''
            },
            address: {
                city: '',
                street: '',
                suite: '',
                zipcode: '',
                geo: {
                    lat: '',
                    lng: ''
                }
            }
        }
    }

    componentWillReceiveProps(nextProps) { 
        this.setState(nextProps.client);
    }

    handleChange(event) {
        switch (event.target.id) {
            case 'name':
                this.setState({name: event.target.value })
                break;
            case 'username':
                this.setState({username: event.target.value })
                break;
            case 'email':
                this.setState({email: event.target.value })
                break;
            case 'phone':
                this.setState({phone: event.target.value })
                break;
            case 'website':
                this.setState({website: event.target.value })
                break;                
        }
        
    }

    doSaveAction = () => {
        this.props.onSave(this.state)
    }

    // componentWillMount() {
    //     const getClient = id =>
    //       fetch(`${API_URL}${API_CLIENT}`+id)
    //         .then(res => res.json())
    //         .catch(err => console.log);
          
    //       from(getClient(this.clientIdToRequest))
    //       .subscribe(client => { 
    //         this.setState({client: client})
    //       })
    //   }

    render() {
        return (
            <form>
                <FormGroup>
                    <ControlLabel>Name</ControlLabel>
                    <FormControl
                        type="text"
                        id="name"
                        value={this.state.name}
                        placeholder="Name"
                        onChange={this.handleChange}
                    />
                    <ControlLabel>Username</ControlLabel>
                    <FormControl
                        type="text"
                        id="username"
                        value={this.state.username}
                        placeholder="Username"
                        onChange={this.handleChange}
                    />
                    <ControlLabel>Email</ControlLabel>
                    <FormControl
                        type="email"
                        id="email"
                        value={this.state.email}
                        placeholder="Email"
                        onChange={this.handleChange}
                    />
                    <ControlLabel>Phone</ControlLabel>
                    <FormControl
                        type="phone"
                        id="phone"
                        value={this.state.phone}
                        placeholder="Phone"
                        onChange={this.handleChange}
                    />
                    <ControlLabel>Website</ControlLabel>
                    <FormControl
                        type="text"
                        id="website"
                        value={this.state.website}
                        placeholder="Website"
                        onChange={this.handleChange}
                    />
                    <Button bsStyle="success" onClick={this.doSaveAction}>Save ðŸ’¾</Button>
                </FormGroup>
            </form>
        )
    }
}

export default ClientForm