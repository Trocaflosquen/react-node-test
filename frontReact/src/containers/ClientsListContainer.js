import React, {Component} from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { fillClientList, removeClient } from '../actions/actionCreators'
import { ListGroupItem, Glyphicon, ButtonToolbar, ButtonGroup, Button } from 'react-bootstrap'
import { ActionLink } from '../components/ActionLink'
import { from } from 'rxjs';
import ClientElement from '../components/ClientElement'
import { withRouter } from "react-router-dom";
import {API_URL, API_CLIENT} from '../config/api-settings'

class ClientsListContainer extends Component {

    constructor(props) {
        super(props)
        this.state = {
            clients: this.props.clients || []
        }
    }

    removeClientAction(id) {
        return () => {
            const removeClientPromise = () =>
            fetch(`${API_URL}${API_CLIENT}`+id, {
                method: 'delete'
              })
                .then(res => res.json())
                .catch(err => console.log);
          
            from(removeClientPromise())
            .subscribe(() => {
                const removedIndex = this.state.clients.findIndex(client => client.id == id)
                this.setState({clients: [...this.state.clients.filter(client => client.id != id)] })
                this.props.removeClient(id)
            })
        } 
    }

    componentWillMount() { 
        const getClients = () =>
          fetch(`${API_URL}/`)
            .then(res => res.json())
            .catch(err => console.log);
          
          from(getClients())
          .subscribe(clients => { 
            this.setState({clients: clients})
            this.props.fillClientList(clients)
          })
      }

    render() { 
        return (this.state.clients && this.state.clients.length > 0) ? (
                this.state.clients.map(
                    client =>
                        (
                            <div className="row" key={client.id}>
                                <div className="col-xs-10">
                                    <ListGroupItem className="client-list-element" onClick={ () => this.props.history.push("/client/"+client.id)}>
                                        <ClientElement {...client}/>
                                    </ListGroupItem>
                                </div>
                                <div className="col-xs-2 list-button">    
                                    <ButtonGroup>
                                        <Button onClick={this.removeClientAction(client.id)}>
                                            ðŸ”ª
                                        </Button>
                                    </ButtonGroup>
                                </div>    
                            </div>
                        )
                    )
        ) : (<p>Loading clients...</p>)
    }
}

const mapStateToProps = state => {
    return {
      clients : state
    }
}
  
const mapDispatchToProps = dispatch => {
    return {
      fillClientList: (clients) => dispatch(fillClientList(clients)),
      removeClient: (clientId) => dispatch(removeClient(clientId))
    }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ClientsListContainer))