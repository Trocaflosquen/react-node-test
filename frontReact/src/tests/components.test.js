import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter as Router } from 'react-router-dom'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import ClientForm from '../components/ClientForm'
import ClientsListContainer from '../containers/ClientsListContainer'
import { clientDataMock, clientArrayMock } from './mocks'

test('ClientForm should be rendered correctly with client data given', () => { 
    const component = renderer.create(
        <ClientForm client={clientDataMock}/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('ClientsListContainer should print a list of clients', () => { 
    const middlewares = []
    const mockStore = configureStore(middlewares)
    const initialState = {}
    const store = mockStore({clients: clientArrayMock})
    
    const component = renderer.create(
        <Provider store={store}>
            <Router>
                <ClientsListContainer />
            </Router>
        </Provider>    ,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});