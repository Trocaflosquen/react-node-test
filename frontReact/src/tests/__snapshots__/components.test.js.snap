// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientForm should be rendered correctly with client data given 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <div>
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Name
      </label>
      <input
        className="form-control"
        defaultValue="Clementine Bauch"
        id="name"
        placeholder="Name"
        type="text"
      />
    </div>
    <div>
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Phone
      </label>
      <input
        className="form-control"
        defaultValue="1-463-123-4447"
        id="phone"
        placeholder="Phone"
        type="text"
      />
    </div>
    <div>
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Username
      </label>
      <input
        className="form-control"
        defaultValue="Samantha"
        id="username"
        placeholder="Username"
        type="text"
      />
    </div>
    <div>
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Email
      </label>
      <input
        className="form-control"
        defaultValue="Nathan@yesenia.net"
        id="email"
        placeholder="Email"
        type="text"
      />
    </div>
    <div>
      <label
        className="control-label"
        htmlFor={undefined}
      >
        Website
      </label>
      <input
        className="form-control"
        defaultValue="ramiro.info"
        id="website"
        placeholder="Website"
        type="text"
      />
    </div>
    <div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          BS
        </label>
        <input
          className="form-control"
          defaultValue="e-enable strategic applications"
          id="bs"
          placeholder="BS"
          type="text"
        />
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Catch Phrase
        </label>
        <input
          className="form-control"
          defaultValue="Face to face bifurcated interface"
          id="catchPhrase"
          placeholder="Catch Phrase"
          type="text"
        />
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Company Name
        </label>
        <input
          className="form-control"
          defaultValue="Romaguera-Jacobson"
          id="name"
          placeholder="Company Name"
          type="text"
        />
      </div>
    </div>
    <div>
      <h3>
        Address
      </h3>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          City
        </label>
        <input
          className="form-control"
          defaultValue="McKenziehaven"
          id="city"
          placeholder="City"
          type="text"
        />
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Street
        </label>
        <input
          className="form-control"
          defaultValue="Douglas Extension"
          id="street"
          placeholder="Street"
          type="text"
        />
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Suite
        </label>
        <input
          className="form-control"
          defaultValue="Suite 847"
          id="suite"
          placeholder="Suite"
          type="text"
        />
      </div>
      <div>
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Zipcode
        </label>
        <input
          className="form-control"
          defaultValue="59590-4157"
          id="zipcode"
          placeholder="Zipcode"
          type="text"
        />
      </div>
    </div>
    <button
      className="btn btn-success"
      disabled={false}
      type="submit"
    >
      Save ðŸ’¾
    </button>
  </div>
</form>
`;

exports[`ClientsListContainer should print a list of clients 1`] = `
<div>
  <div
    className="static-modal"
  />
  <div
    className="main-buttons btn-toolbar"
    role="toolbar"
  >
    <button
      className="btn btn-success"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add client âž•
    </button>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-10"
    >
      <button
        className="client-list-element list-group-item"
        onClick={[Function]}
        type="button"
      >
        <p>
          ðŸ‘¤ 
          Leanne Graham
        </p>
      </button>
    </div>
    <div
      className="col-xs-2 list-button"
    >
      <div
        className="btn-group"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          ðŸ”ª
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-10"
    >
      <button
        className="client-list-element list-group-item"
        onClick={[Function]}
        type="button"
      >
        <p>
          ðŸ‘¤ 
          Ervin Howell
        </p>
      </button>
    </div>
    <div
      className="col-xs-2 list-button"
    >
      <div
        className="btn-group"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          ðŸ”ª
        </button>
      </div>
    </div>
  </div>
</div>
`;
